### ISSUE TITLE
Implementation Reports

### ISSUE THREAD

**Issue by otherdaniel** [WICG/sanitizer-api#139]
This isn't a bug; it's just a place to collect experience with using the Sanitizer API.

The goal is to have a single place to gather usage feedback, and to make it easy for everyone to drop us a line or two about how it went for them. By collecting these in one place, it'll hopefully be easier to see if there're any themes that we could/should address in the API.

**Comment by otherdaniel** [comment]
**Trial Re-Implementation of the Angular Sanitizer on top of Sanitizer API**

[Angular](https://github.com/angular/angular) contains a [built-in HTML sanitizer](https://github.com/angular/angular/tree/master/packages/core/src/sanitization). As an experiment, we re-implemented this sanitizer based on the Sanitizer API, but retained the API shape and security properties of the original.

Observations:

**API shape**: The Angular sanitizer is string-to-string. It doesn't support custom configuration options, presumably with the goal of being simple and safe. Both of these are easily accommodated, although the former requires the non-recommended use of `.sanitizeFor(...).innerHTML`.

We found only one substantial **difference in the security model** that can't be modelled by the Sanitizer API configuration: Angular Sanitizer removes javascript:-URLs in non-navigation contexts. (E.g. `<img src="javascript:bla">`.) This used to cause script execution in ancient browsers, but doesn't do so in any contemporary browser we know of. This is a value-dependent sanitization action and hence cannot be configured. Emulating this by post-processing the DOM tree (before string-ifying it with `.innerHTML`) was easy. (From our perspective, this was purely a compatibility, rather than a security measure.)

**Secure Context restriction**: This gave us a surprisingly hard time. Our original goal was to quickly hack together an implementation for our trial and to then run it over the substantial test suites for several in-house Angular applications, in order to find out how it holds up in real world usage, and whether there's any "blind spots" in the API and/or our implementation. It turns out that tests have a strong tendency to set up a minimal environment for testing. And since the "original" in this case would work just fine in non-secure `http:`, few of the relevant tests bothered to setup a secure test server. This meant that testing would have required us to re-factor several test suites first. We opted to not do this part.
